File: 416-partition-equal-subset-sum/partition-equal-subset-sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = 0;
        for (int i: nums)
            sum += i;
        int target = sum / 2;
        if ((sum % 2) == 1) return false;

        unordered_set<int> mapper;
        mapper.insert(0);

        for (int i = nums.size() - 1; i >= 0; i--){
            unordered_set<int> temp;
            for (int j: mapper){
                if (j + nums[i] == target) return true;
                temp.insert(j + nums[i]);
                temp.insert(j);
            }
            mapper = temp;
        }
        return false;
    }
};

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        if sum(nums)%2 == 1: return False

        mapper = set()
        mapper.add(0)
        target = sum(nums)//2
        for i in range(len(nums) - 1, -1, -1):
            temp = set()
            for j in mapper:
                if nums[i] + j == target:
                    return True
                temp.add(nums[i] + j)
                temp.add(j)
            mapper = temp
        return False
